File: Day-07_No-Space-Left-On-Device/cpp/main.cpp
#include <bits/stdc++.h>

struct File
{
    std::string name;
    std::size_t size;
};

class Directory
{
public:
    Directory* parent;
    std::vector<Directory*> subdirs;
    std::vector<File> files;
    std::string name;
    std::string path;

    Directory(const std::string &name=""): name(name)
    {
        path = name + std::string("/");
    }

    void addSubdir(const std::string &name)
    {
        Directory* dir = new Directory(name);
        dir->parent = this;
        dir->path = this->path + dir->name + std::string("/");
        subdirs.push_back(dir);
    }

    void addFile(const File& file)
    {
        files.push_back(file);
    }

    void printFilesystem(std::size_t tabs=0)
    {
        std::cout << std::string(tabs, '\t') << this->name << "/" << std::endl;
        for(auto &i : subdirs)
        {
            i->printFilesystem(tabs+1);
        }
        for(auto &i : files)
        {
            std::cout << std::string(tabs + 1, '\t') << i.name << " " << i.size << std::endl;
        }
    }

    std::size_t getFileSize()
    {
        std::size_t totalSize = 0;
        for(auto &i : subdirs)
        {
            totalSize += i->getFileSize();
        }
        for(auto &i : files)
        {
            totalSize += i.size;
        }
        return totalSize;
    }

    std::size_t getPartOneAnswer()
    {
        std::size_t total = 0;
        for(auto &i : subdirs)
        {
            total += i->getPartOneAnswer();
        }
        std::size_t fsize = getFileSize();
        if(fsize < 100000) total += fsize;
        return total;
    }

    // I know directories are not files but file datatype suits just fine
    std::vector<File> getPartTwoCandidates(std::size_t neededSize)
    {
        std::vector<File> result;
        for(auto &i : subdirs)
        {
            auto b = i->getPartTwoCandidates(neededSize);
            result.insert(result.end(), b.begin(), b.end());
        }
        for(auto &i : subdirs)
        {
            std::size_t fsize = i->getFileSize();
            if(fsize >= neededSize)
            {
                result.push_back({i->name, fsize});
            }
        }
        return result;
    }

    void clearDirs()
    {
        for(auto &i : subdirs)
        {
            i->clearDirs();
            delete i;
        }
    }
};

std::vector<std::string> split(const std::string &str, char delim)
{
    std::string currentStr = std::string();
    std::vector<std::string> result;

    for(std::size_t i = 0; i < str.size(); i++)
    {
        if(str[i] == delim)
        {
            result.push_back(currentStr);
            currentStr = std::string();
        }
        else
        {
            currentStr += str[i];
        }
    }
    result.push_back(currentStr);
    return result;
}

void changeDir(Directory **dirptr, const std::string &name)
{
    if(name == "..")
    {
        if(dirptr != nullptr) *dirptr = (*dirptr)->parent;
        return;
    }
    for(auto &s : (*dirptr)->subdirs)
    {
        if(s->name == name)
        {
            *dirptr = s;
            break;
        }
    }
}

int main()
{
    // std::shared_ptr<Directory> dirPtr = std::make_shared<Directory>("Parent");
    Directory parentDir("");
    Directory *dirPtr = &parentDir;

    std::vector<std::string> inputs;
    while(true)
    {
        std::string input;
        std::getline(std::cin, input);
        if(input.length() == 0) break;
        inputs.push_back(input);
    }

    for(auto &input : inputs)
    {
        auto splitted = split(input, ' ');
        if(splitted[0] == "$")
        {
            if(splitted[1] == "cd")
            {
                if(splitted[2] == "/") continue;
                changeDir(&dirPtr, splitted[2]);
            }
        }
        else if(splitted[0] == "dir")
        {
            dirPtr->addSubdir(splitted[1]);
        }
        else
        {
            std::size_t fsize = std::atoi(splitted[0].c_str());
            std::string fname = splitted[1];
            dirPtr->addFile({fname, fsize});
        }
    }

    std::cout << "Part One: " << parentDir.getPartOneAnswer() << std::endl;

    const std::size_t maxSpace = 70000000;
    std::size_t freeSpace = maxSpace - parentDir.getFileSize();
    const std::size_t reqSpace = 30000000;
    std::size_t neededSpace = reqSpace - freeSpace;

    std::size_t smallestSize = -1;
    for(auto &i : parentDir.getPartTwoCandidates(neededSpace))
    {
        // std::cout << i.name << ": " << i.size << std::endl;
        if(i.size < smallestSize)
        {
            smallestSize = i.size;
        }
    }

    std::cout << "Part Two: " << smallestSize << std::endl;

    parentDir.clearDirs();
    return 0;
}

File: Day-07_No-Space-Left-On-Device/src/main.rs
use std::{io, ptr};

// This code do not work. I tried a lot of times and eventualy solved this problem with c++.
// My mistake was not declaring variables on heap.
// I tried allocating but I get a lot of errors

#[derive(Clone)]
pub struct File{
    pub size: usize,
    pub name: String,
}

#[derive(Clone)]
pub struct Directory{
    pub parent: *mut Box<Directory>,
    pub subdirs: Vec<*mut Box<Directory>>,
    pub files: Vec<File>,
    pub name: String,
    pub path: String,
}

impl Directory{
    fn new(dirName: &str) -> Box<Directory>{
        return Box::new(Directory{
            parent: ptr::null_mut(),
            subdirs: Vec::new(),
            files: Vec::new(),
            name: dirName.to_string().clone(),
            path: format!("{}{}", dirName, "/"),
        });
    }

    unsafe fn addSubdir(&mut self, newSubDir: &mut Box<Directory>){
        for &dir in &self.subdirs{
            if (*dir).name == newSubDir.name{
                return;
            }
        }
        self.subdirs.push(newSubDir);
        newSubDir.parent = self;
        newSubDir.path = format!("{}{}", self.path, newSubDir.name);
        if newSubDir.path.chars().nth(newSubDir.path.len()-1).unwrap() != '/'{
            newSubDir.path.push('/');
        }
    }

    fn addFile(&mut self, filename: &str, filesize: usize){
        self.files.push(File{
            name: filename.to_string(),
            size: filesize,
        });
    }

    fn getFolderSize(&self) -> usize{
        let mut totalSize: usize = 0;
        for &d in &self.subdirs{
            unsafe{totalSize += (*d).getFolderSize();};
        }

        for f in &self.files{
            totalSize += f.size;
        }

        return totalSize;
    }

    fn printFilesystem(&mut self, tabCount: usize){
        println!("{}{}/", "\t".repeat(tabCount), self.name);
        for &d in &self.subdirs{
            unsafe {(*d).printFilesystem(tabCount + 1);};
        }
        for f in &self.files{
            println!("{}{} {}", "\t".repeat(tabCount + 1), f.name, f.size);
        }
    }
}

unsafe fn changeDir(dirptr: &mut *mut Directory, dirname: &str){
    if dirname == ".."{
        if (*(*dirptr)).parent.is_null(){
            return;
        }
        // I am completely lost
        (*dirptr) = (*(*dirptr)).parent;
        return;
    }
    for &d in &(*(*dirptr)).subdirs{
        unsafe{
            if dirname.to_string() == (*d).name {
                (*dirptr) = &mut (*d);
                break;
            }
        };
    }
}
//
// static mut parentFolder: Directory = Directory::new("");
// static mut currentDir: *mut Directory = &mut unsafe { parentFolder };

fn main() {
    let mut inputs: Vec<String> = Vec::new();

    loop {
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("");
        input = input[0..input.len() - 1].to_string();
        if input.is_empty() {
            break;
        }
        inputs.push(input);
    }

    let mut parentDir: Directory = Directory::new("Parent");
    let mut currentDir: *mut Directory = &mut parentDir;
    let mut allDirs: Vec<Directory> = Vec::new();

    let mut somedir: Directory = Directory::new("Somedir");
    unsafe { parentDir.addSubdir(&mut Box::new(Directory::new("asd"))); }

    unsafe{
        for input in &inputs{
            let splitted: Vec<&str> = input.split(" ").collect();
            if splitted[0] == "$"{
                if splitted[1] == "cd"{
                    if splitted[2] == "/"{
                        continue;
                    }
                    println!("Changedir {} to {}", (*currentDir).name, splitted[2]);
                    changeDir(&mut currentDir, splitted[2]);
                }
            }
            else if splitted[0] == "dir"{
                println!("Adddir: {:?}", splitted);
                // somedir = Directory::new(splitted[1]);
                // (*currentDir).addSubdir(&mut somedir);
                parentDir.addSubdir(&mut *Box::new(Directory::new(splitted[1])))
            }
            else{
                println!("Addfile");
                let fsize: usize = splitted[0].parse().unwrap();
                let fname = splitted[1];
                (*currentDir).addFile(fname, fsize);
            }
        }
    };

    parentDir.printFilesystem(0);
    for d in &allDirs{
        println!("{}", d.name);
    }

    // unsafe {
    //     (*currentDir).printFilesystem(0);
    //     changeDir(&mut currentDir, "Child");
    //     (*currentDir).printFilesystem(0);
    //     changeDir(&mut currentDir, "..");
    //     (*currentDir).printFilesystem(0);
    //
    // };
}
